File: Python/08 다이나믹 프로그래밍/12865_평범한 배낭.py
import sys
input = sys.stdin.readline

n, k  = map(int, input().split())

bag = []
for _ in range(n):
    bag.append(list(map(int, input().split())))

dp = [[0] * (k+1) for _ in range(n+1)]

for i in range(1, n+1):
    for j in range(1, k+1):     # 최대 무게까지
        if j >= bag[i-1][0]:    # 현재 최대 무게가 해당 물건무게보다 크면

            # 현재 물건 가치 + 이전 물건 가치의 최대값
            dp[i][j] = max(bag[i-1][1] + dp[i-1][j - bag[i-1][0]], dp[i-1][j])
        else:
            dp[i][j] = dp[i-1][j]

print(dp[n][k])

File: JavaScript/Dynamic Programming/12865_평범한배낭.js
const fs = require("fs");
const input = fs.readFileSync("/dev/stdin").toString().trim().split("\n");

const [N, K] = input[0].split(" ").map(Number);

const items = input.slice(1).map((line) => {
  const [w, v] = line.split(" ").map(Number);
  return { w, v };
});

let dp = Array(K + 1).fill(0);

items.forEach(({ w, v }) => {
  for (let weight = K; weight >= w; weight--) {
    dp[weight] = Math.max(dp[weight], dp[weight - w] + v);
  }
});

console.log(dp[K]);
